connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/WorkerCoordinator.java

```diff
@@ -145,154 +166,79 @@ public void poll(long timeout) {
     }
 
     @Override
-    public JoinGroupRequestData.JoinGroupRequestProtocolSet metadata() {
+    public JoinGroupRequestProtocolSet metadata() {
         configSnapshot = configStorage.snapshot();
-        ConnectProtocol.WorkerState workerState = new ConnectProtocol.WorkerState(restUrl, configSnapshot.offset());
-        ByteBuffer metadata = ConnectProtocol.serializeMetadata(workerState);
-        return new JoinGroupRequestData.JoinGroupRequestProtocolSet(
-                Collections.singleton(new JoinGroupRequestData.JoinGroupRequestProtocol()
-                        .setName(DEFAULT_SUBPROTOCOL)
-                        .setMetadata(metadata.array()))
-                        .iterator()
-        );
+        ExtendedWorkerState workerState = new ExtendedWorkerState(restUrl, configSnapshot.offset(), assignmentSnapshot);
+        ByteBuffer metadata;
+        switch (protocolCompatibility) {
```

Comment:
Right, and that would be easier to unit test, too.

Reply:
Dynamic dispatch is not very easy or elegant here, because the protocol classes are effectively static classes (have only static methods) and we can't hold on to a reference on their instance and use inheritance effectively. I've moved the creation of `JoinGroupRequestProtocolCollection` in the protocols though, and now the `switch` block (which I also dislike compared to dynamic dispatch) is small.